File: WordBreak_BottomUp.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        '''
        Bottom Up Approach
        Time: O(n^3)
        Space: O(n)
        '''
        x = set(wordDict)
        dp = [False for i in range(0,len(s))]
        # iterate over case where s = "leetcode"
        for i in range(0,len(s)):
            for j in range(0,i+1):
                if(j>0):
                    # check if leet is true
                    if(dp[j-1]==True):
                        # if yes, check code is in x
                        if(s[j:i+1] in x):
                            dp[i] = True
                            break
                else:
                    # this is mainly require to check for j=0, we need not to check dp[j-1] is true
                    if(s[j:i+1] in x):
                        dp[i] = True
                        break
        
        return dp[-1]

File: WordBreakBottomUp.java
import java.util.ArrayList;
import java.util.List;
import java.util.HashSet;
import java.util.Scanner;

public class WordBreakBottomUp {

        // BOTTOM-UP DYNAMIC PROGRAMMING - TIME O(N*M*L) + O(M*L) AND SPACE O(N) + O(M*L)

        public boolean wordBreak(String s, List<String> wordDict) {

            // store word dictionary in a set for checking substrings
            HashSet<String> set = new HashSet<>(wordDict);     // O(m*l)

            boolean[] dp = new boolean[s.length()+1];

            dp[0] = true;                              // O(n) space

            // build bottom up dp array
            for(int i = 1; i < dp.length; i++) {      // O(n)

                // if dp[j] is true, previous sub string is valid. So check rest of sub string till iterator-i
                for(int j = 0; j < i; j++) {         // O(m*l)

                    if(dp[j] && set.contains(s.substring(j, i))) {

                        dp[i] = true;

                        break;
                    }
                }
            }
            //
            return dp[s.length()];
        }

        public static void main(String[] args) {

            WordBreakBottomUp obj = new WordBreakBottomUp();

            Scanner scanner = new Scanner(System.in);

            System.out.print("string s: ");
            String s = scanner.nextLine();

            System.out.print("word dictionary size: ");
            int l = scanner.nextInt();
            scanner.nextLine();

            System.out.print("word dictionary: ");
            List<String> wordDict = new ArrayList<>();

            for (int i = 0; i < l; i++) {

                String word = scanner.nextLine();
                wordDict.add(word);
            }

            boolean answer = obj.wordBreak(s, wordDict);

            System.out.print("Can string s be segmented into a space-separated sequence of one " +
                    "or more dictionary words? " + answer);
        }

}


/*
TIME COMPLEXITY = O(N*M*L) + O(M*L)

SPACE COMPLEXITY = O(N) + O(M*L)
*/